### PR TITLE
Fix build warnings and update compiler

### PR DESCRIPTION
**Pull request by sharwell**

### CODE REVIEW

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidUninstantiatedInternalClasses.cs [hunk 1]
```diff
@@ -44,7 +44,7 @@ public override void Initialize(AnalysisContext analysisContext)
 
             analysisContext.RegisterCompilationStartAction(startContext =>
             {
-                var instantiatedTypes = new ConcurrentDictionary<INamedTypeSymbol, object?>();
+                ConcurrentDictionary<INamedTypeSymbol, object?> instantiatedTypes = new ConcurrentDictionary<INamedTypeSymbol, object?>();
```
Comment (jcouv) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidUninstantiatedInternalClasses.cs]: Those explicit types were not necessary, are they?

Comment (Evangelink) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidUninstantiatedInternalClasses.cs]: The commit introducing these changes is named "Use explicit type where compiler fails to evaluate nullability" so I think they matters.

Comment (sharwell) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidUninstantiatedInternalClasses.cs]: @jcouv yes, they are necessary. `var` doesn't work with local functions, at least not here.

Comment (mavasani) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidUninstantiatedInternalClasses.cs]: Do we need to file a compiler bug?

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotPassLiteralsAsLocalizedParameters.cs [hunk 2]
```diff
@@ -66,7 +66,7 @@ public override void Initialize(AnalysisContext context)
                         return;
                     }
 
-                    var lazyValueContentResult = new Lazy<DataFlowAnalysisResult<ValueContentBlockAnalysisResult, ValueContentAbstractValue>?>(
+                    Lazy<DataFlowAnalysisResult<ValueContentBlockAnalysisResult, ValueContentAbstractValue>?> lazyValueContentResult = new Lazy<DataFlowAnalysisResult<ValueContentBlockAnalysisResult, ValueContentAbstractValue>?>(
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotPassLiteralsAsLocalizedParameters.cs]: Same here - this seems unfortunate.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Security/DoNotReferSelfInSerializableClass.cs [hunk 3]
```diff
@@ -66,14 +66,14 @@ public sealed override void Initialize(AnalysisContext context)
                         return;
                     }
 
-                    var forwardGraph = new ConcurrentDictionary<ISymbol, ConcurrentDictionary<ISymbol, bool>>();
-                    var invertedGraph = new ConcurrentDictionary<ISymbol, ConcurrentDictionary<ISymbol, bool>>();
+                    ConcurrentDictionary<ISymbol, ConcurrentDictionary<ISymbol, bool>> forwardGraph = new ConcurrentDictionary<ISymbol, ConcurrentDictionary<ISymbol, bool>>();
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Security/DoNotReferSelfInSerializableClass.cs]: Please retarget all security rule changes to 2.9.x branch. Also, there is no nullability involved here, so why was this required?

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidUninstantiatedInternalClasses.cs (hunk 1 above)
Comment (sharwell) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidUninstantiatedInternalClasses.cs]: I think there are already bugs filed for improving local function support. It hit Roslyn too.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotPassLiteralsAsLocalizedParameters.cs (hunk 2 above)
Comment (sharwell) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotPassLiteralsAsLocalizedParameters.cs]: ➡️ Covered by https://github.com/dotnet/roslyn-analyzers/pull/3400#discussion_r394759842. The compiler requires this change.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Security/DoNotReferSelfInSerializableClass.cs (hunk 3 above)
Comment (sharwell) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Security/DoNotReferSelfInSerializableClass.cs]: It's not practical to backport this change. The 2.9.x branch doesn't support this toolchain or compiler version so it would be a large backport of much more than just this pull request.

I don't believe these changes will be a problem in practice. /cc @dotpaul for awareness.

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidUninstantiatedInternalClasses.cs (hunk 1 above)
Comment (sharwell) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidUninstantiatedInternalClasses.cs]: ➡️ https://github.com/dotnet/roslyn/issues/41689 and https://github.com/dotnet/roslyn/issues/41705
